Administrator User Case , Domain Model

minseungseon opened the pull request:
# Task

resolves #5 

# 추가 작업

- Domain Model 을 UC 10, 11 추가하기 
- 그림 그리기 diagram! 

# 생각해봐야할 것

- UC 10과 11 은 다른 관련 UC 들을 generalization 을 해놓은 Use Case 인데, 이것에 대해서 Domain Model 을 명세해야되는지.
(교수님께 질문완료, Generalization된 Use case에 대해서는 Domain Model이 작성될 필요가 없다. 대신, 지금 다른 Use Case와 Domain Model이 잘 작성된 것인지 확인 필요)
- 관리자에서 request 를 할 때 add pizza(POST), edit pizza(PATCH, PUT) 가 정보 입력과정이 같은데 add, edit pizza 의 겹치는 부분은 공통으로 쓰고 다른 부분만 parameter 로 줘서 domain model 에서 구분을 해야하는건가? 
(다시 고민을 해보았는데, 각각의 Use Case에 대해서만 Domain Model이 작성되면 되므로 고려할 필요가 없을 것 같습니다.)

Topslug committed:
Create Administrator.md

Topslug committed:
Create administrator_use_case.md

Topslug committed:
Update README.md

Topslug committed:
Update administrator_use_case.md

Topslug committed:
Update Administrator.md

Use Case를 분석하는 도중 옵션을 선택하는 것을 특정한 Use Case로 두기가 애매하다는 생각이 들었었는데, Functional Requirement보다는 UI Requirement에 가깝다는 생각이 들어서 변경했습니다.

Topslug committed:
Update administrator_use_case.md

Topslug committed:
Update administrator_use_case.md

Topslug committed:
Update administrator_use_case.md

Topslug committed:
Update administrator_use_case.md

Topslug committed:
Add files via upload

Topslug committed:
Delete ViewingRecordsSystem.png

Topslug committed:
Add files via upload

Topslug committed:
Update administrator_use_case.md

Topslug committed:
Create administrator_domain_model.md

Topslug committed:
Update administrator_domain_model.md

Topslug committed:
Update administrator_domain_model.md

Topslug committed:
Delete EmployeeManagementSystem.png

Topslug committed:
Delete PizzaManagementSystem.png

Topslug committed:
Add files via upload

minseungseon reviewed (commented):
1차 검토 했습니다!

minseungseon commented on adminisrator/administrator_domain_model.md:
```diff
@@ -0,0 +1,317 @@
+| 새로운 Pizza에 대한 정보 입력 요청                  |  K   | AddRequest       |
+| Pizza에 대한 정보를 입력할 수 있도록하는 HTML 문서  |  K   | InputPage |
+| InputPage에 대한 HTML 문서를 생성한다.              |  D   | PageMaker           |
+| 추가된 Pizza가 반영된 새로운 피자관리페이지 HTML문서 | K    | UpdatedPage        |
+| 추가된 Pizza가 반영된 새로운 피자관리페이지 HTML문서를 생성한다.    | D    | PageMaker    |
```
해결!

minseungseon commented on Administrator.md:
```diff
@@ -2,12 +2,13 @@
+7. 관리자는 직원을 추가, 수정, 삭제 할 수 있어야 한다. 
+8. '관리자' 권한을 부여할 수 있는 기능이 필요하고, 권한이 있다면 위의 2~7번을 실행할 수 있게 된다.
 
+### User Interface Requirements
+1. 관리자는 옵션을 선택하여 자신이 원하는 작업을 수행할 수 있어야 한다.ex("Add Pizza", "Edit Pizza", “View Order Records”, “View Feedback”, "Manage Employee" 등의 옵션 중 선택하여 작업을 수행할 수 있어야 한다.)
```
이 부분은 UI function 에 가깝기 때문에, functional requirements에 해당하지 않다고 생각해서 따로 뺐다!

Topslug committed:
Add files via upload

minseungseon reviewed (commented):
상민씨와 회의하며 리뷰 한번더 완료!

Topslug committed:
Update Administrator.md

Topslug reviewed (commented):
Use Case - 1 로그인과정에서 로그인에 대한 정보를 데이터베이스와 상호작용하도록 작성했는데 다른 분들의 생각은 시스템내에 정보가 있는 것으로 본 것 같습니다. 수정하도록 하겠습니다.

Topslug committed:
Create OrderInformation.md

Topslug committed:
Create CustomerInformation.md

Topslug committed:
Update Administrator.md

경준님의 의견을 반영하였습니다.
생각해보니 직원이 관리자 권한을 획득하는 경우 관리자 또한 직원의 일을 할 수 있어야 하므로 Use Case의 수정이 필요할 것 같습니다.

Topslug committed:
Update administrator_use_case.md

Topslug committed:
Create Employe_display.md

Topslug committed:
Create 새로운형식의시나리오.md

Topslug commented on adminisrator/새로운형식의시나리오.md:
```diff
@@ -0,0 +1,11 @@
+관리자는 시스템을 총괄하는 피자가게 사장님입니다. 
+하지만 관리자는 직원처럼 일하기도 합니다. 
+직원처럼 피자를 만들기도 하고 주문을 처리하기도 합니다. 
+또한 직원을 고용하기도 하고, 해고하기도 합니다. 
+일 잘하는 직원의 월급은 올려주기도 합니다. 
+새로운 피자를 개발하기도 하고, 아주 잘팔리는 피자에 대해서는 가격을 올려팔기도 합니다. 
+잘 팔리지 않는 피자의 경우 판매 목록에서 지워버리기도 합니다. 
+관리자는 항상 피자를 어떻게 하면 더 많이 팔 수 있을지 고민합니다. 
+그래서 매일매일 지금까지 얼마나 많은 주문이 들어왔는지 보고는 합니다. 
+고객들은 피자에 Review를 남겨주기도 하는데 그것을 바탕으로 피자개발에 참고 하기도 합니다. 
+관리자는 종종 가게를 비우기도 하는데 그럴때에는 가장 믿음직한 직원에게 가게관리를 맡기기도 합니다.
```
지금까지는 사장님이 주문 처리를 못하는 것으로 되있었는데,
경준님의 회원 권한 변경에 관련된 의견과, 직원 권한을 관리자 권한으로 바꿀 수 있다는 기능으로 봤을 때,
관리자도 직원처럼 주문 처리를 할 수 있게 하는 것이 맞는 것 같아서 이렇게 수정하였습니다.
그런데 지금 형식은 이전에 썼던 방식과는 다른 형식인데, 경준님께서 이전 시나리오가 이런 형식으로 쓰여지는 것이 맞는가?라는 의문을 제시하셨었는데, 그 의견을 참고해서 작성해보았습니다.

Topslug committed:
Update 새로운형식의시나리오.md

Topslug committed:
Update 새로운형식의시나리오.md

Topslug committed:
Update administrator_use_case.md

회원가입에 관한 사항 추가

Topslug committed:
Update administrator_use_case.md

회원가입에 관한 사항 추가

Topslug reviewed (commented):
생각해보니 Use Case - View Pizza Information, Use Case - View Employee Information도 추가할 수 있을 것 같습니다.
그러면 Use case상에서 include표현을 통해 좀 더 간략한 표현이 가능할 것 같습니다.

추가적으로 Generalization Use Case : View information도 가능할 것 같습니다.

minseungseon commented on adminisrator/administrator_domain_model.md:
```diff
@@ -0,0 +1,317 @@
+| StoreRequest ↔ Controller       | Controller는 새로운 피자정보 저장에 대한 요청을 전달받는다     | receives        |
+| StoreRequest ↔ BlankChecker     | StoreRequest는 BlankChecker에게 input parameter를 전달한다.   | receives        |
+| Controller ↔ BlankChecker       | Controller는 input parameter에 빈칸이 있는지 물어본다.        | checks blank    |
+| PageMaker ↔ Warning             | PageMaker는 Warning을 만든다.                                 | prepares        |  
+| Controller ↔ DatabaseConnection | Controller는 DatabaseConnection에게 입력된 피자정보저장 요쳥을 전달한다.       | conveys request  |
```
Page Maker <-> Warning 중복되어있습니다!! (26번줄과요!~~)

minseungseon commented on adminisrator/administrator_domain_model.md:
```diff
@@ -0,0 +1,317 @@
+| Controller ↔ PageMaker          | Controller는 PageMaker에게 SelectPage 또는 ReviewPage를 만들라고 요청한다 그리고 만들어진 page를 받는다. | conveys request |
+| PageMaker ↔ SelectPage          | PageMaker는 SelectPage를 만든다.                             | prepares         |
+| PageMaker  ↔ RecordPage         | PageMaker는 ReviewPage를 만든다.                             | prepares        |
+| Controller ↔ RecordPage         | Controller는 SelectPage를 사용자에게 전달한다.                | posts           |
+| Controller ↔ RecordPage         | Controller는 ReviewPage를 사용자에게 전달한다.                | posts           |
```
같은 pair 에 대해서 두개가 존재하고, 각각은 RecordPage 안에서 다른 역할을 명시하기 위해서 중복되게 해놓은건가요?

Topslug replied:
아.. 이 부분은 제가 실수를 한 것 같습니다. 죄송합니다. 수정이 필요할 것 같습니다.

minseungseon reviewed (commented):
리뷰 추가해두었습니다! :)

minseungseon committed:
Create DomainModelDiagram정리본.md

minseungseon committed:
Update DomainModelDiagram정리본.md

minseungseon committed:
[docs] UC2 설명글 추가

minseungseon committed:
[upload] UC3 diagram 추가

minseungseon committed:
[refac ] UC3 - edit pizza 다이어그램 수정

minseungseon committed:
[upload] UC4 delete pizza 다이어그램 추가

minseungseon committed:
[upload] UC 5 - view order records 다이어그램 추가

minseungseon committed:
[upload] UC 6 - view reviews diagram 추가

minseungseon committed:
[upload] UC 7 - add employee 다이어그램 추가

minseungseon committed:
[upload] UC8 , 9 diagram 추가

minseungseon committed:
[refac] 상민씨에게 전달 사항 추가

Topslug commented:
만약 매장이 여러 개여서 가맹점에 서비스하는 프로그램으로 할 경우
Pizza Menu와 Customer를 관리하는 최고 관리자가 필요해 질 것 같습니다. 팀과의 전체 회의를 통해 시스템 구성을 정해야 될 것 같은데 만약 최고 관리자가 필요해지게 되는 경우
Actor로 Major Admin, Minor Admin을 추가하여
Admin으로 Generalization을 할 수 있을 것 같습니다.
또한 Major Admin으로 접근하는 경우에만 Manage Pizza와 Manage Customer를 할 수 있도록 하고, Major Admin, Minor Admin 모두 공통적으로 해당하는 인터페이스는 Manage Employee, View Order Records, View Reviews, Manage Current Orders가 될 것 같습니다.

그리고 생각해보니 관리자 본인의 계정을 설정할 수 있는 부분을 추가하지 않아서 이 부분에 대한 추가가 필요할 것 같습니다.
또  저는 직원 계정을 하나로 통일하는 방안으로 생각 중인데
Use Case 관리자 계정 설정
Use Case 직원 계정 설정
Generalization Use Case 계정 설정
이런 식으로 일반화할 수 있지 않을까 싶습니다.

카톡 상으로 말씀 드렸던 내용이긴 한데
사장님도 주문을 처리할 수 있도록 하는 과정이 필요할 것 같습니다.
하지만 이 부분은 Employee팀과 토의해서
Employee, Admin 
Generalization Worker
이런 식으로 일반화해볼 수 있지 않을까 싶습니다.

정리 (앞으로 하면 좋을 것들)
- 계정 설정에 관한 Use Case를 만든다.
- Major Admin과 Minor Admin을 Admin으로 일반화하여 표현하고 그에 맞게 Use Case를 수정한다.
- 관리자도 직원처럼 주문 처리를 할 수 있도록 하는 과정이 필요할 것 같은데 Employee팀과 상의 후 결정이 필요해 보인다.

Topslug commented on adminisrator/DomainModelDiagram정리본.md:
```diff
@@ -0,0 +1,37 @@
+# DomainModelDiagram 정리본 공유
```
깔끔하게 잘 정리하신 것 같습니다! 아직 자세히 보지는 못했지만 같이 확인하면서 점검해봐요! :)

minseungseon replied:
넵

Topslug commented:
Database에 대해서 전체 시스템 구성이 정해지면 수정이 필요할 것 같습니다.

수정이 필요할 것 같은 부분
- Diagram 상에서 database삭제 가능성
- Deriving 상에서 database 삭제 가능성
- Domain Model에서 해당 Database 삭제 가능성

Topslug commented:
일을 자꾸 만드는 것 같아 죄송합니다..
그런데 로그인 기능이 있으면 로그아웃 기능이 있어야 될 것 같은 생각이 자꾸 들어서..
그리고 직원 관리 페이지에서 관리자 페이지
리뷰 목록 페이지에서 관리자 페이지로 가는 상황 등에 대해서도 토의를 해봐야 될 것 같습니다.
Requirement에 저희가 추가하지는 못했지만, 관리자는 본인이 원하는 옵션을 선택해서 자유롭게 작업이 가능하다..? 이 부분에 포함 되는 것 같기도 합니다..

Topslug committed:
Create administrator_use_case2.md

View information 관련 Use Case 추가
Requirement에 회원가입, 계정 설정 부분 추가

Topslug committed:
Update administrator_use_case2.md

Topslug committed:
Update administrator_use_case2.md

Topslug committed:
Create administrator_domain_model2.md

Topslug committed:
Update administrator_use_case2.md

Topslug committed:
Update administrator_use_case2.md

Topslug committed:
Update administrator_domain_model2.md

Topslug committed:
Add files via upload

Topslug committed:
Update administrator_use_case2.md

Topslug committed:
Update administrator_use_case2.md

Topslug committed:
Update administrator_use_case2.md

Topslug committed:
Update administrator_use_case2.md

Topslug committed:
Update administrator_use_case2.md

Topslug committed:
Update administrator_domain_model2.md

Topslug committed:
Delete Domain model 12~19.pdf

Topslug committed:
Add files via upload

Topslug committed:
Delete Domain model 2.pdf

Topslug committed:
Add files via upload

Topslug committed:
Update administrator_use_case2.md

Topslug committed:
Delete ViewingRecordsSystem.png

Topslug committed:
Delete PizzaManagementSystem.png

Topslug committed:
Delete OrderInformation.md

Topslug committed:
Delete Employe_display.md

Topslug committed:
Delete EmployeeManagementSystem.png

Topslug committed:
Update administrator_use_case2.md

Topslug committed:
Delete CustomerInformation.md

Topslug committed:
Add files via upload

minseungseon commented:
뭔가 한번은 만나서 같이 얘기해봐야할것 같네용 🤔🤔🤔🤔🤔

minseungseon commented:
일단 actor 들 추가한 부분 추가해보겠습니다! 
그리고 저희 지금 파일이 너무 분산되어있어서 정리해봐야할 것 같아요...!

minseungseon committed:
[upload] Domain Model Diagram 수정본 업로드

Topslug committed:
Update administrator_use_case2.md

minseungseon committed:
[upload] domain model diagram 상민씨 부분 추가
